src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -370,9 +456,46 @@ void onDeviceOnOff() {
     }
 
     void onKnob(int index, double normalized) {
-      if (this.knobs[index] != null) {
-        this.knobs[index].setNormalized(normalized);
+      LXListenableParameter knob = this.knobs[index];
+
+      if (knob == null) {
+        return;
+      }
+
+      // If the knob's an LCP, touching it should link it to the focusColor,
+      // if one is set and it belongs to the main palette and not a side
+      // swatch, or set it static and initialize it to the colorClipboard,
+      // if there is one. Otherwise, have it control the value as normal.
+      if (knob instanceof LinkedColorParameter) {
+        LinkedColorParameter lcp = (LinkedColorParameter) knob;
+
+        if (focusColor != null) {
+          int palIndex = lx.engine.palette.swatch.colors.indexOf(focusColor);
+          if (palIndex >= 0) {
+            lcp.mode.setValue(LinkedColorParameter.Mode.PALETTE);
+            lcp.index.setValue(palIndex + 1);
+            return;
+          }
+        }
+        if (colorClipboard != null) {
+          lcp.mode.setValue(LinkedColorParameter.Mode.STATIC);
+          lcp.setColor(colorClipboard);
+          return;
+        }
+      }
+
+      LXListenableNormalizedParameter knobParam = parameterForKnob(knob);
+
+      // If it's a wrappable parameter, let it wrap
+      if (knobParam.isWrappable()) {
+        if (normalized == 0.0) {
+          normalized = 1.0;
+        } else {
+          normalized = (normalized + 1.0) % 1.0;
```

Comment:
This is sort of confusing to read, I think might be clearer if you had an `else if (normalized == 1.0)` set it to 0. With a comment, that would make clear that this is catching the cases where the knob has just spun UP to 1., or DOWN to 0.

This mod-math expression *only* actually results in a different value for `1.0`, which gets changed to `0.0` - but this is not immediately clear from looking at it. I think parity with the above expression turning 0.0 to 1.0 will make this more understandable.

Reply:
Oh, I didn't realize that. Done ... but it's too bad that by the time onKnob() is called, the value is already normalized (truncated) because it means if the current value is 0.99 and they give it a big turn that would normally be worth an adjustment of +0.25, the new value is 0.0 instead of 0.24.